enhancements/console/dynamic-plugins.md

```diff
@@ -0,0 +1,369 @@
+
+### Shared Dependencies
+
+Console will expose React and Patternfly 4 as shared dependencies to plugins.
+Plugins can use any Patternfly component. Only a single version of React and
+Patternfly will be loaded.
+
+Plugins will not be able to use legacy Patternfly 3 components.
+
+### Delivering Plugins
+
+Plugins are delivered through operators. The operators will typically create a
+deployment on the platform with an HTTP server to host the plugin and expose it
+using a Kubernetes `Service`. The `Service` must use HTTPS and a
+[service serving certificate](https://docs.openshift.com/container-platform/4.4/security/certificates/service-serving-certificate.html).
+The console backend will load the plugin directly from the `Service` using the
+service CA bundle.
+
+Operators declare that they have a console plugin available by creating a
+`ConsolePlugin` resource that includes the service name, port, and path to the
+plugin manifest.
+
+```yaml
+apiVersion: console.openshift.io/v1alpha1
+kind: ConsolePlugin
+metadata:
+  name: my-plugin
+spec:
+  displayName: My Plugin
+  service:
+    name: my-console-plugin
+    namespace: my-operator-namespace
+    port: 8443
+    manifest: /manifest.json
+```
+
+Plugins are disabled by default. They need to be manually enabled by a cluster
+administrator before console loads any plugin code. Console provides a UI
+for enabling plugins, and the list of enabled plugins is set in the console
+operator config.
+
+```yaml
+apiVersion: operator.openshift.io/v1
+kind: Console
+metadata:
+  name: cluster
+spec:
+  managementState: Managed
+  plugins:
+  - my-plugin
+```
+
+This will trigger a new rollout of the console `Deployment` with the updated
+plugins. Note that this does mean the user will need to refresh the browser
+to see the new plugins.
+
+An operator can also give console a hint that it has a plugin available through
+the `console.openshift.io/plugins` annotations. The annotation tells the console to
+show a checkbox to enable plugins during operator install. The value is a
+serialized JSON array of strings, where each item is the name of a
+`ConsolePlugin` resource the operator will create.
+
+```yaml
+apiVersion: operators.coreos.com/v1alpha1
+kind: ClusterServiceVersion
+metadata:
+  annotations:
+    console.openshift.io/plugins: '["my-plugin"]'
+```
+
+For operators from the `redhat-operators` catalog source, the install operator
+page will check the plugin checkbox by default. For other operators, the
+checkbox will not be checked by default. Any plugin can be enabled or disabled
+later through the console cluster settings page.
+
+### Loading Plugins
+
+The console backend will proxy requests to the plugin `Service`. This way the
+`Service` itself does not need to be exposed outside the cluster, and avoids
+any same-origin or certificate errors in the browser loading JavaScript from a
+separate domain. 
+
+On load, console will read the manifest for each plugin. The manifest is a JSON
```

Comment:
on what load?  on browser load?  or on console backend startup?

if it's on backend startup, how/when does this content get refreshed, such as when the CNV operator is upgraded and updates its console plugin bits?